x-pack/plugins/observability_solution/observability_onboarding/public/application/packages_list/index.tsx

```diff
@@ -0,0 +1,201 @@
+  setSearchQuery,
+}: WrapperProps) => {
+  const [isInitialHidden, setIsInitialHidden] = useState(showSearchBar);
+  const customMargin = useCustomMargin();
+  const availablePackages = useAvailablePackages({
+    prereleaseIntegrationsEnabled: false,
+  });
+  const { filteredCards } = availablePackages;
+  const list: IntegrationCardItem[] = useIntegrationCardList(
+    filteredCards,
+    selectedCategory,
+    featuredCardNames,
+    generatedCards
+  );
+  React.useEffect(() => {
+    if (isInitialHidden && searchQuery) {
+      setIsInitialHidden(false);
+    }
+  }, [searchQuery, isInitialHidden]);
+  if (!isInitialHidden && availablePackages.isLoading) return <Loading />;
+
+  const showPackageList =
+    (showSearchBar && !isInitialHidden) || showSearchBar === false;
+
+  return (
+    <Suspense fallback={<Loading />}>
+      <div css={customMargin}>
+        {showSearchBar && (
+          <div
+            css={css`
+              max-width: 600px;
+            `}
+          >
+            <EuiSearchBar
+              box={{
+                incremental: true,
+                inputRef: (ref: any) => {
+                  (
+                    searchBarRef as React.MutableRefObject<HTMLInputElement>
+                  ).current = ref;
+                },
+              }}
+              onChange={(arg) => {
+                if (setSearchQuery) {
+                  setSearchQuery(arg.queryText);
+                }
+                setIsInitialHidden(false);
+              }}
+              query={searchQuery}
+            />
+          </div>
+        )}
+        {showPackageList && (
+          <PackageList
+            list={list}
+            searchTerm={searchQuery ?? ''}
+            showControls={false}
+            showSearchTools={false}
+            // we either don't need these properties (yet) or handle them upstream, but
+            // they are marked as required in the original API.
+            selectedCategory={selectedCategory}
+            setSearchTerm={() => {}}
+            setCategory={() => {}}
+            categories={[]}
+            setUrlandReplaceHistory={() => {}}
+            setUrlandPushHistory={() => {}}
+          />
+        )}
+      </div>
+    </Suspense>
+  );
+};
+
+const WithAvailablePackages = React.forwardRef(
+  (props: Props, searchBarRef?: React.Ref<HTMLInputElement>) => {
+    const ref = useRef<AvailablePackagesHookType | null>(null);
+    const [reloadRetry, setReloadRetry] = useState(0);
+    const [errorLoading, setErrorLoading] = useState(false);
+    const [loadingModule, setLoadingModule] = React.useState(true);
+
+    useEffect(() => {
+      async function load() {
+        setErrorLoading(false);
+        setLoadingModule(true);
+        try {
+          ref.current = await fetchAvailablePackagesHook();
+        } catch (e) {
+          setErrorLoading(true);
+        } finally {
+          setLoadingModule(false);
+        }
+      }
+      load();
+    }, [reloadRetry]);
```

Comment:
I would suggest using a hook like `useAsyncRetry` from `react-use` package for loading and error states which handles all the complexity around it for you: https://streamich.github.io/react-use/?path=/story/side-effects-useasyncretry--docs

Trying to implement this manually often causes issues for weird edge cases. For example an async function could resolve when React has already unmounted the component causing an error. Your manual logic here would not handle this scenario correctly.

Reply:
TIL about `useAsyncRetry` - I really need to sit down and read all the docs for `react-use` sometime. So many useful shortcuts.

I will change this ++

Reply:
So much cleaner https://github.com/elastic/kibana/pull/179301/commits/c2dfb88cb92cbfb471aa316f515610ce473f4449